lib/components/user/phone-verification-form.tsx

```diff
@@ -0,0 +1,149 @@
+})
+
+interface Fields {
+  validationCode: string
+}
+
+export type PhoneVerificationSubmitHandler = (values: Fields) => void
+
+interface Props {
+  isSubmitting: boolean
+  onRequestCode: PhoneChangeSubmitHandler
+  onSubmit: PhoneVerificationSubmitHandler
+}
+
+type InnerProps = Props & FormikProps<Fields>
+
+/**
+ * Sub-form for entering and submitting a phone validation code.
+ */
+const PhoneVerificationInnerForm = (props: InnerProps): JSX.Element => {
+  const { errors, isSubmitting, onRequestCode, submitCount, touched } = props
+  const ref = useRef<HTMLInputElement>()
+
+  // Return focus to the input control if the incorrect code was submitted.
+  useEffect(() => {
+    if (!isSubmitting && submitCount) {
+      ref.current?.focus()
+    }
+  }, [isSubmitting, submitCount])
+
+  const codeErrorState = getErrorStates(props).validationCode
+  const isInvalid = !!(touched.validationCode && errors.validationCode)
+  const formId = 'phone-verification-form'
+
+  return (
+    <FormGroup validationState={codeErrorState}>
+      {/* Set up an empty Formik Form without inputs, and link inputs using the form id.
+                    (A submit button within will incorrectly submit the entire page instead of just the subform.)
+                    The containing Formik element will watch submission of the form. */}
+      <Form id={formId} noValidate />
+      <p>
+        <FormattedMessage id="components.PhoneNumberEditor.verificationInstructions" />
+      </p>
+      <ControlLabel htmlFor="validation-code">
+        <FormattedMessage id="components.PhoneNumberEditor.verificationCode" />
+      </ControlLabel>
+      <ControlStrip>
+        <Field
+          aria-invalid={isInvalid}
+          aria-required
+          as={InlineTextInput}
+          form={formId}
+          id="validation-code"
+          maxLength={6}
+          name="validationCode"
+          placeholder="123456"
+          ref={ref}
+          // HACK: <input type='tel'> triggers the numerical keypad on mobile devices, and otherwise
+          // behaves like <input type='text'> with support of leading zeros and the maxLength prop.
+          // <input type='number'> causes values to be stored as Number, resulting in
+          // leading zeros to be invalid and stripped upon submission.
+          type="tel"
+        />
+        <Button
+          bsStyle="primary"
+          disabled={isSubmitting}
+          form={formId}
+          type="submit"
+        >
+          {isSubmitting ? (
+            <InlineLoading />
+          ) : (
+            <FormattedMessage id="components.PhoneNumberEditor.verify" />
+          )}
+          <InvisibleA11yLabel role="status">
+            {isSubmitting && <FormattedMessage id="common.forms.submitting" />}
+          </InvisibleA11yLabel>
+        </Button>
+        <HelpBlock role="alert">
+          {isInvalid && (
+            <FormattedMessage id="components.PhoneNumberEditor.invalidCode" />
```

Comment:
Basically, the form only checks that you are entering digits, but it can't obviously tell whether the code was correct. That is handled by the phone verification service after the form is submitted. We're just clearing up the field if the submitted code was wrong.

Reply:
Good catch. This one was hard to fix, see https://github.com/opentripplanner/otp-react-redux/pull/799/commits/fd0f9ab8e62c82435014d77bc612edf0fcc0462c.

Reply:
Works great @binh-dam-ibigroup! Thanks for this change